### PR TITLE
Implement spatial awareness scaffolding

### PR DESCRIPTION
**Pull request by davidkline-ms**
Overview
---
Add a skeletal implementation for the spatial awareness abstraction layer - underlying functionality to come in a future PR (to keep this one as manageable as possible).

Impemented the spatial awareness profile and inspector (hooked into the core configuration profile)

Update architecture documentation to match changes identified while creating the scaffolding.

Fix a minor bug or two in the configuration profile.

Fixes: #2508, #2386 

Created: #2724 to track Windows Mixed Reality implementation

### CODE REVIEW

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/Handlers/MixedRealitySpatialAwarenessSurfaceFindingHandler.cs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/Handlers/MixedRealitySpatialAwarenessSurfaceFindingHandler.cs]: no namespace?

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessBaseDescription.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        int IEqualityComparer.GetHashCode(object obj)
+        {
+            return obj.GetHashCode();
+        }
+
+        public override int GetHashCode()
+        {
+            string s = $"Position({Position.x},{Position.y},{Position.z})";
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessBaseDescription.cs]: I think you can just use `Position` as is correct?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessBaseDescription.cs]: Unity's string overload is pretty print

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessMeshDescription.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+namespace Microsoft.MixedReality.Toolkit.SDK.SpatialAwarenessSystem
+{
+    /// <summary>
+    /// Class poviding the default implementation of the <see cref="IMixedRealitySpatialAwarenessMeshDescription"/> interface.
+    /// </summary>
+    public class MixedRealitySpatialAwarenessMeshDescription : MixedRealitySpatialAwarenessBaseDescription, IMixedRealitySpatialAwarenessMeshDescription
+    {
+        /// <inheritdoc />
+        public Mesh Mesh
+        {get; private set;}
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessMeshDescription.cs]: same line?

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+
+            IsObserverRunning = true;
+        }
+
+        /// <inheritdoc />
+        public void SuspendObserver()
+        {
+            if (!IsObserverRunning) { return; }
+
+#if UNITY_WSA
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs]: We shouldn't have any platform specific code in our main system, but in platform components.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs]: Similar to how we did devices for input system.

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+            get
+            {
+                return meshTrianglesPerCubicMeter;
+            }
+
+            set
+            {
+                if (IsObserverRunning)
+                {
+                    throw new System.InvalidOperationException("MeshTrianglesPerCubicMeter cannot be modified while the observer is running.");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs]: Let's try to keep throwing exceptions to a minimun as they crash the player/editor. Let's raise an error instead.

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+        public bool RenderPlatformSurfaces { get; set; } = false;
+
+        /// <inheritdoc />
+        public Material PlatformSurfaceMaterial { get; set; } = null;
+
+        /// <inheritdoc />
+        public Dictionary<uint, IMixedRealitySpatialAwarenessPlanarSurfaceDescription> GetSurfaces()
+        {
+            // todo
+            return new Dictionary<uint, IMixedRealitySpatialAwarenessPlanarSurfaceDescription>(0);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs]: Do we wanna make a new dictionary every time, or modify an existing dictionary (as not all the surfaces may have changed)

File: Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityConfigurationProfile.asset [hunk 7]
```diff
@@ -11,11 +11,7 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: 7612acbc1a4a4ed0afa5f4ccbe42bee4, type: 3}
   m_Name: DefaultMixedRealityConfigurationProfile
   m_EditorClassIdentifier: 
-  initialManagerTypes:
-  - reference: Microsoft.MixedReality.Toolkit.SDK.Input.MixedRealityInputManager,
-      Microsoft.MixedReality.Toolkit.SDK
-  - reference: Microsoft.MixedReality.Toolkit.SDK.BoundarySystem.MixedRealityBoundaryManager,
-      Microsoft.MixedReality.Toolkit.SDK
+  initialManagerTypes: []
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityConfigurationProfile.asset]: We seem to be getting a lot of chern on this line.

File: Assets/MixedRealityToolkit/_Core/Resources/Shaders/MRTK_Wireframe.shader [hunk 8]
```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+///
+/// Basic wireframe shader that can be used for rendering spatial mapping meshes.
+///
+Shader "Mixed Reality Toolkit/Wireframe"
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Resources/Shaders/MRTK_Wireframe.shader]: Can't we just use the one built into the editor resources?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealitySpatialAwarenessProfileInspector.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            if (!CheckMixedRealityManager())
+            {
+                return;
+            }
+
+            if (GUILayout.Button("Back to Configuration Profile"))
+            {
+                Selection.activeObject = MixedRealityManager.Instance.ActiveProfile;
+            }
+
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealitySpatialAwarenessProfileInspector.cs]: Don't forget to add the button to return to the main config profile

File: Assets/MixedRealityToolkit/_Core/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessMeshEventData.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        public IMixedRealitySpatialAwarenessMeshDescription MeshDescription { get; private set; }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="eventSystem"></param>
+        public MixedRealitySpatialAwarenessMeshEventData(EventSystem eventSystem) : base(eventSystem) { }
+
+        public void Intialize(
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessMeshEventData.cs]: missing interit doc summary

File: Assets/MixedRealityToolkit/_Core/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessSurfaceFindingEventData.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        public IMixedRealitySpatialAwarenessPlanarSurfaceDescription SurfaceDescription { get; private set; }
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="eventSystem"></param>
+        public MixedRealitySpatialAwarenessSurfaceFindingEventData(EventSystem eventSystem) : base(eventSystem) { }
+
+        public void Intialize(
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessSurfaceFindingEventData.cs]: missing interit doc summary

File: Assets/MixedRealityToolkit/_Core/Definitions/SpatialAwareness/IMixedRealitySpatialAwarenessPlanarSurfaceDescription.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Describes the data required for an application to understand how to construct and place a planar surface in the environment.
+    /// </summary>
+    public interface IMixedRealitySpatialAwarenessPlanarSurfaceDescription : IMixedRealitySpatialAwarenessBaseDescription
+    {
+        /// <summary>
+        /// The axis aligned bounding box that contains the surface being described.
+        /// </summary>
+        Bounds BoundingBox { get; }
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/SpatialAwareness/IMixedRealitySpatialAwarenessPlanarSurfaceDescription.cs]: I think people might get this confused with a UX bounding box.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs [hunk 13]
```diff
@@ -252,6 +260,17 @@ public override void OnInspectorGUI()
                 EditorGUILayout.PropertyField(teleportDuration);
             }
 
+            // Spatial Awareness System configuration
+            GUILayout.Space(12f);
+            EditorGUILayout.LabelField("Spatial Awareness System Settings", EditorStyles.boldLabel);
+            EditorGUILayout.PropertyField(enableSpatialAwarenessSystem);
+
+            if (enableSpatialAwarenessSystem.boolValue)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs]: Please remove this so we don't hide it when disabled.

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/Handlers/MixedRealitySpatialAwarenessSurfaceFindingHandler.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/Handlers/MixedRealitySpatialAwarenessSurfaceFindingHandler.cs]: odd.... not sure how that happened... i'll fix

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessBaseDescription.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessBaseDescription.cs]: cool. i'll do that. thx!

File: Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityConfigurationProfile.asset (hunk 7 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityConfigurationProfile.asset]: strange... no idea why that would be

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityConfigurationProfile.asset]: I think it has something to do with the MR Manager's serialization of the currently running managers. No related, but just thought it was curious.

File: Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs (hunk 4 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit-SDK/Features/SpatialAwareness/System/MixedRealitySpatialAwarenessSystem.cs]: we likely will have to have something to trigger the correct observer. this is purely placeholder for now

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealitySpatialAwarenessProfileInspector.cs (hunk 9 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealitySpatialAwarenessProfileInspector.cs]: i took this code from the boundary profile as is... i see the button and have tested that it works.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs (hunk 13 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityConfigurationProfileInspector.cs]: every system in the core profile behave this way

File: Assets/MixedRealityToolkit/_Core/Definitions/SpatialAwareness/IMixedRealitySpatialAwarenessPlanarSurfaceDescription.cs (hunk 12 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/SpatialAwareness/IMixedRealitySpatialAwarenessPlanarSurfaceDescription.cs]: Bounds? 😅

**Pull request merged**